src/protocol/boolean/generate_random_bits/malicious.rs

```diff
@@ -0,0 +1,57 @@
+use super::{RandomBits, Step};
+use crate::error::Error;
+use crate::ff::{Field, Int};
+use crate::protocol::context::MaliciousContext;
+use crate::protocol::modulus_conversion::{convert_bit, convert_bit_local};
+use crate::protocol::{context::Context, BitOpStep, RecordId};
+use crate::secret_sharing::{MaliciousReplicated, XorReplicated};
+use async_trait::async_trait;
+use futures::future::try_join_all;
+use std::iter::repeat;
+
+#[async_trait]
+impl<F: Field> RandomBits<F> for MaliciousContext<'_, F> {
+    type Share = MaliciousReplicated<F>;
+
+    /// Generates a sequence of `l` random bit sharings in the target field `F`.
+    async fn generate_random_bits(self, record_id: RecordId) -> Result<Vec<Self::Share>, Error> {
+        // Calculate the number of bits we need to form a random number that
+        // has the same number of bits as the prime.
+        let l = u128::BITS - F::PRIME.into().leading_zeros();
+        let leading_zero_bits = F::Integer::BITS - l;
+
+        // Generate a pair of random numbers. We'll use these numbers as
+        // the source of `l`-bit long uniformly random sequence of bits.
+        let (b_bits_left, b_bits_right) = self
+            .narrow(&Step::RandomValues)
+            .prss()
+            .generate_values(record_id);
+
+        // Same here. For now, 256-bit is enough for our F_p
+        let xor_share = XorReplicated::new(
+            u64::try_from(b_bits_left & u128::from(u64::MAX)).unwrap(),
+            u64::try_from(b_bits_right & u128::from(u64::MAX)).unwrap(),
+        );
+
+        // Convert each bit to secret sharings of that bit in the target field
+        let c = self.narrow(&Step::ConvertShares);
+        let futures = (0..l).map(|i| {
+            let c = c.narrow(&BitOpStep::from(i));
+            let triple = convert_bit_local::<F>(c.role(), i, &xor_share);
+            async move {
+                let malicious_triple = c.upgrade_bit_triple(record_id, i, triple).await?;
```

Comment:
This is the only line different from the semi-honest version.

Reply:
I do not love this level of code duplication. How about you split the semi-honest version into two separate function calls:
1. beginning of this function up until it calls `convert_bit_local`
2. a function that accepts "triples", calls "convert_bit" on them, then appends the leading zeros.

Then the malicious version can just call function #1, call `upgrade_bit_triple`, then call function #2. 
The semi-honest version will just call function #1 followed by function #2.

Reply:
Fixed as suggested.